### PR TITLE
AP_Frsky_Telem: Squash inits into one

### PR DESCRIPTION
**Pull request by floaledm**
Open to comments on how to better check for null references, but tested working on all three vehicles...

### CODE REVIEW

File: ArduPlane/system.cpp [hunk 1]
```diff
@@ -162,7 +162,7 @@ void Plane::init_ardupilot()
     // setup frsky, and pass a number of parameters to the library
     frsky_telemetry.init(serial_manager, FIRMWARE_STRING,
                          MAV_TYPE_FIXED_WING,
-                         &g.fs_batt_voltage, &g.fs_batt_mah);
+                         &g.fs_batt_voltage, &g.fs_batt_mah, 0);
```
Comment (magicrub) [inline ArduPlane/system.cpp]: lets use
&g.fs_batt_voltage, &g.fs_batt_mah, nullptr);

File: APMrover2/system.cpp [hunk 2]
```diff
@@ -129,7 +129,7 @@ void Rover::init_ardupilot()
 
     // setup frsky telemetry
 #if FRSKY_TELEM_ENABLED == ENABLED
-    frsky_telemetry.init(serial_manager);
+    frsky_telemetry.init(serial_manager, FIRMWARE_STRING, MAV_TYPE_GROUND_ROVER, 0, 0, 0);
```
Comment (magicrub) [inline APMrover2/system.cpp]: lets use
frsky_telemetry.init(serial_manager, FIRMWARE_STRING, MAV_TYPE_GROUND_ROVER, nullptr, nullptr, nullptr);

File: libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp [hunk 3]
```diff
@@ -47,9 +46,15 @@ void AP_Frsky_Telem::init(const AP_SerialManager &serial_manager, const char *fi
         queue_message(MAV_SEVERITY_INFO, firmware_str);
         // save main parameters locally
         _params.mav_type = mav_type; // frame type (see MAV_TYPE in Mavlink definition file common.h)
-        _params.fs_batt_voltage = fs_batt_voltage; // failsafe battery voltage in volts
-        _params.fs_batt_mah = fs_batt_mah; // failsafe reserve capacity in mAh
-        _ap.value = ap_value; // ap bit-field
+        if (fs_batt_voltage) {
```
Comment (magicrub) [inline libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp]: lets use:
if (fs_batt_voltage != nullptr) {

**Comment by floaledm** [comment]
Made the changes and also added else {*_ap.value = 0;}. If you don't think that's necessary, I can certainly remove it...

File: libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp [hunk 4]
```diff
@@ -47,9 +46,17 @@ void AP_Frsky_Telem::init(const AP_SerialManager &serial_manager, const char *fi
...
+        if (fs_batt_voltage != nullptr) {
+            _params.fs_batt_voltage = fs_batt_voltage; // failsafe battery voltage in volts
+        }
+        if (fs_batt_mah != nullptr) {
+            _params.fs_batt_mah = fs_batt_mah; // failsafe reserve capacity in mAh
+        }
+        if (ap_value != nullptr) {
+            _ap.value = ap_value; // ap bit-field
+        } else {
+            *_ap.value = 0;
```
Comment (OXINARF) [inline libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp]: This is wrong, you are assigning a value to an unknown location (in the case of ArduPilot the location is known to be 0, but you should never do this).

Also what is the need for all the ifs? If it receives nullptr there's no harm in assigning nullptr...

File: ArduPlane/system.cpp [hunk 5]
```diff
@@ -162,7 +162,7 @@ void Plane::init_ardupilot()
     // setup frsky, and pass a number of parameters to the library
     frsky_telemetry.init(serial_manager, FIRMWARE_STRING,
                          MAV_TYPE_FIXED_WING,
-                         &g.fs_batt_voltage, &g.fs_batt_mah);
+                         &g.fs_batt_voltage, &g.fs_batt_mah, nullptr);
```
Comment (OXINARF) [inline ArduPlane/system.cpp]: If it is always the last parameters that can be nullptr, then declare the init function with nullptr for those, then you don't need to pass nullptr arguments.

**Comment by magicrub** [comment]
Good ideas. Just assign the nullptr then do the check later in the code
when it gets used.

**Comment by magicrub** [comment]
@OXINARF and @floaledm please take a look at my branch here that has a few proposed changes:
https://github.com/magicrub/ardupilot/commits/pr/squash-inits

**Comment by OXINARF** [comment]
@magicrub Looks good except the initialization in the header. Not only that should be done by default but it will also be done by the constructor.

**Comment by magicrub** [comment]
>except the initialization in the header

Whats wrong with it as-is? I don't understand why the constructor needs it, that would just be redudant

**Comment by OXINARF** [comment]
I'm not an expert in that, but, as far as I know, it increases the binary size - I'm sure you know we never want initialization in headers unless it is needed, and there isn't a need to initialize something to nullptr, it is already constructed that way. Also, what I said about the constructor is that it already initializes to nullptr.

**Comment by magicrub** [comment]
Well, I was setting to nullptr to be explicit instead of zero

**Comment by magicrub** [comment]
But yeah, it could be removed..

**Comment by magicrub** [comment]
merged, thanks @OXINARF and @floaledm !

**Comment by floaledm** [comment]
Thanks @magicrub and @OXINARF  for all your work! I was away for the weekend...

**Comment by floaledm** [comment]
There's an issue with ap_value on Plane which I'll address shortly in another PR...

**Pull request closed without merging**